presto-tests/src/test/java/com/facebook/presto/tests/TestQueryTaskLimit.java

```diff
@@ -90,16 +90,31 @@ public void testExceedTaskLimit()
     public void testQueuingWhenTaskLimitExceeds()
             throws Exception
     {
-        try (DistributedQueryRunner queryRunner = createQueryRunner(defaultSession, ImmutableMap.of())) {
-            QueryId firstQuery = createQuery(queryRunner, newSession("test", ImmutableSet.of(), null), LONG_LASTING_QUERY);
+        ImmutableMap<String, String> extraProperties = ImmutableMap.<String, String>builder()
+                .put("experimental.spill-enabled", "false")
+                .put("experimental.max-total-running-task-count-to-not-execute-new-query", "2")
+                .build();
+
+        try (DistributedQueryRunner queryRunner = createQueryRunner(defaultSession, extraProperties)) {
+            QueryId firstQuery = createQuery(queryRunner, newSession("test", ImmutableSet.of(), null),
+                    LONG_LASTING_QUERY);
             waitForQueryState(queryRunner, firstQuery, RUNNING);
 
-            queryRunner.getCoordinator().getResourceGroupManager().get().setTaskLimitExceeded(true);
+            // wait for a while, the first query will schedule at lease 2 tasks, so exceed resource group manager's total task limit
+            try {
+                Thread.sleep(500);
+            }
+            catch (Exception e) {
+                // do noting
+            }
 
             QueryId secondQuery = createQuery(queryRunner, newSession("test", ImmutableSet.of(), null), LONG_LASTING_QUERY);
+
+            // When current running tasks exceeded limit, the following query would be queued
             waitForQueryState(queryRunner, secondQuery, QUEUED);
 
-            queryRunner.getCoordinator().getResourceGroupManager().get().setTaskLimitExceeded(false);
+            // When first query is cancelled, the second query would be pass to run
+            queryRunner.getCoordinator().getDispatchManager().cancelQuery(firstQuery);
```

Comment:
Wait for cancellation of first query before proceeding to check next query?

Reply:
The logic to check next query is a while loop to wait for desired state. As the first query did not finish it's cancellation, the second query would continuously in QUEUED state, so function waitForQueryState(queryRunner, secondQuery, RUNNING) would continuously loop util the first query finish it's cancellation and the second query start to run. So, here we can call waitForQueryState(queryRunner, secondQuery, RUNNING) immediately, and no need to wait.

Reply:
Yes, implicitly it's the same, but it would be nice to check that the query is cancelled explicitly (rather than timed out for example).

Reply:
OK, I get your point. You are right, It's better to check explicitly the first query is cancelled. I will fix it.